guides/inventing-service.md

```diff
@@ -0,0 +1,866 @@
+            if request.path() == "/" {
+                Ok(Response::ok("Hello, World!"))
+            } else if request.path() == "/important-data" {
+                let some_data = fetch_data_from_database().await?;
+                Ok(make_response(some_data))
+            } else {
+                Ok(Response::not_found())
+            }
+        })
+    }
+}
+```
+
+How about supporting timeouts? Remember, the solution we're aiming for is one
+that allows us to combine different pieces of functionality together without
+having to modify each individual piece.
+
+What if we defined a generic `Timeout` struct like this:
+
+```rust
+struct Timeout<T> {
+    // T will be some type that implements `Handler`
+    inner_handler: T,
+    duration: Duration,
+}
+```
+
+We could then implement `Handler` for `Timeout<T>` and delegate to `T`'s
+`Handler` implementation:
+
+```rust
+impl<T> Handler for Timeout<T>
+where
+    T: Handler,
+{
+    type Future = Pin<Box<dyn Future<Output = Result<Response, Error>>>>;
+
+    fn call(&mut self, request: Request) -> Self::Future {
+        Box::pin(async move {
+            let result = tokio::time::timeout(
+                self.duration,
+                self.inner_handler.call(request),
+            ).await;
+
+            match result {
+                Ok(Ok(response)) => Ok(response),
+                Ok(Err(error)) => Err(error),
+                Err(_timeout) => Err(Error::timeout()),
+            }
+        })
+    }
+}
+```
+
+The important line being `self.inner_handler.call(request)`. This is where we
+delegate to the inner handler and let it do its thing. We don't know what it is
+we just know that it produces a `Result<Response, Error>` when its done.
+
+This code doesn't quite compile though. We get an error like this:
+
+```
+error[E0759]: `self` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
+   --> src/lib.rs:145:29
+    |
+144 |       fn call(&mut self, request: Request) -> Self::Future {
+    |               --------- this data with an anonymous lifetime `'_`...
+145 |           Box::pin(async move {
+    |  _____________________________^
+146 | |             let result = tokio::time::timeout(
+147 | |                 self.duration,
+148 | |                 self.inner_handler.call(request),
+...   |
+155 | |             }
+156 | |         })
+    | |_________^ ...is captured here, requiring it to live as long as `'static`
+```
+
+The issue is that we're capturing a `&mut self` and moving it into an async
+block. That means the lifetime of our future is tied to the lifetime of `&mut
+self`. This doesn't work for us since we might want to run our response futures
+on multiple threads to get better performance or produce multiple response
+futures and run them all in parallel. That isn't possible if a reference to the
+handler lives inside the futures.
+
+Instead we need convert the `&mut self` into an owned `self`. That is exactly
```

Comment:
another option is because the inner `Handle::Future` _has_ to be `'static` you can:

```
let fut = self.inner.call(...);
```

and that `fut` has to be a `'static` future.

Reply:
The clone method also works though :D